### ISSUE TITLE
Not closing cURL

### ISSUE THREAD

**Issue by fattalgazi** [liamcottle/Instagram-SDK-PHP#28]
Hello,

I have a php class that includes your SDK, it is working in infinite loop, after a couple of hours later it consumes a lot of opened file limit in the system, I investigated the situation and found that you never close curl connections. Please look this commit:

https://github.com/liamcottle/Instagram-SDK-PHP/commit/cc7575c8e0da0db5964d217fdf56dede724b3dce

is it ok ? Thanks.

**Comment by SteveJobzniak** [comment]
@fattalgazi @liamcottle You are 100% correct about the cURL handles, fattalgazi. It's a very, very severe and nasty bug in this library.

The code leaves **every** cURL handle/resource in memory after a request is complete (even successful requests). It steals RAM and steals open file handles. And after X amount of API queries (new cURL handles), you will get a FATAL PHP error: "failed to open stream: Too many open files" and the PHP script will die.

@liamcottle If you have Unix (Mac) or Linux, you can easily verify this bug for yourself by setting a VERY low file handle limit for the PHP process. Type "unlimit -n 12" in a Terminal window. That sets your maximum number of file handles for that bash session to 12 open files at a time. Then type "php <somescript>" to run any script that performs more than 12 Instagram API queries. Your script will soon die.

But with @fattalgazi's patch, the script will live forever even with just 12 open handles as the max limit. Personally, I actually encountered this bug on a system a handle limit of 256, which is insane.

However, fattalgazi's fix isn't clean, and I am not sure every codepath is covered. But all credit goes to him for discovering the source of the problem!

To clean it up, I rewrote Framework/Request.php to use a try{}finally{} block. The finally{} is _always_ executed even if the code in the try-block throws an exception or returns. So it's great for releasing resources at the end of function. The "finally{}" feature requires PHP 5.5 (http://php.net/manual/en/language.exceptions.php), but I highly recommend it, because it is the de-facto correct programming method for guaranteed release of resources.

Here is my finally{}-patch. It looks like a lot of changes, but it's really just the try{}finally{}, with indentation of the old code to move it into the try{}.

View: https://gist.github.com/SteveJobzniak/8bbdeb6c4e231db1fb5032385b70aecd

Raw: https://gist.githubusercontent.com/SteveJobzniak/8bbdeb6c4e231db1fb5032385b70aecd/raw/a73313085f7412b82128126f302896602d7ca87d/releaseCurlResources.patch

This code can be merged as-is or changed as you see fit, @liamcottle. It 100% guarantees that the cURL resources are released 100% of the time no matter what happens.

One thing is for sure: This _needs_ to be patched. My way or some other way.

My way works, but another (perhaps even better) option might be to reuse a single cURL handle internally? It seems a bit wasteful to "new Curl()" every time. ;)

**Comment by liamcottle** [comment]
Hiya,

Yes, this needs to be fixed ... However, I can't really tell what was changed by looking at the patch file due to the indentation changes.

I'm going to be rewriting this library on top of the [Guzzle PHP library](https://github.com/guzzle/guzzle) which will solve these issues, and make the code of this library a lot cleaner.

I don't have an ETA on this at this stage, as I'm currently working on other things.

If you can submit this as a pull request, I can test a merge.

Cheers !

**Comment by liamcottle** [comment]
Heh, fully missed this part

>It looks like a lot of changes, but it's really just the try{}finally{}, with indentation of the old code to move it into the try{}.

Yeah I see now 😅

A pull request would be appreciated ! Cheers.

**Comment by SteveJobzniak** [comment]
@liamcottle Here you go! :) Equally ugly diff on GitHub but now in pull request form:

https://github.com/liamcottle/Instagram-SDK-PHP/pull/29

I've now been running this patch for an hour with the file handle limit set to 25, and have zero leaks. I can conclude that this is now fixed. The original code crashed within a minute. The fattal code crashed within a few minutes. And this one simply always releases the handle and stays chugging forever. Choo choo!

About Guzzle: Yeah it kicks ass. I use Guzzle all the time!

**Comment by liamcottle** [comment]
Sweet! Merging now. I'll need to update composer.json to require PHP 5.5+

**Comment by fattalgazi** [comment]
You are awesome @SteveJobzniak and @liamcottle ! 
Actually, the problem was not negligible as you think, on my production server "ulimit -n" is set to "65536" as quite enough. After running 3 php files in infinite loop which includes unpatched version of this class, and requesting to instagram 20-30k times, each process consumed nearly 20k open file limit.  (you can see which pid consumes how much open file limit by : lsof -n  | awk '{print $1,$2}'  | uniq -c | sort -nr | head -100 )

Long story short, after 1 hour, there was no open file limit left on the system and every process started to be killed because of fatal errors. It looks like a small patch but believe me only 3 php processes knocked out entire server. Thanks for your attention and fixes.
